### PR TITLE
Jetpack connect: Use promises for createAccount

### PR DESCRIPTION
**Pull request by sirreal**
The end goal is to return a promise from the action that can be consumed in the component.
There's not reason to go through Redux state for these cases where a component needs to perform an API call and proceed depending on the result.

By consuming a promise in the component we should be able to reduce significant Redux complexity.

Related to #23483 

## Testing
1. Create a new Jurassic Ninja site
1. Must be logged out of WordPress.com
1. Follow the connection -> sign up flow for the new site
1. Signup and connection should work normally

## Error testing
1. Account creation errors are not handled well currently. This PR does not change that and some limitations (signup form cannot change state to `submitting=false`, not global notices in logged out #23951) prevent improvements. I'll tackle that in a follow-up
1. To force an API error, I've added D11559-code to make error testing simple. Apply it, sandbox `public-api.wordpress.com` and uncomment one of the errors. Verify that master fails in the same state as this branch: dead form, no notices 😬 
1. You _can_ verify that this branch has the pieces in place for better error handling. The `submitting` state and prop on the `SignupForm` should be `false` after the error, and a `notice` action should have been dispatched:
   ```js
   actionLog.filter(/NOTICE/)
   ```

### CODE REVIEW

File: client/jetpack-connect/signup.js [hunk 1]
```diff
@@ -76,31 +77,35 @@ export class JetpackSignup extends Component {
...
-		this.props.createAccount( userData );
+		this.props.createAccount( userData ).then(
+			bearerToken =>
+				this.setState( {
+					newUsername: userData.username,
+					bearerToken,
+				} ),
+			error => {
+				debug( 'Signup error: %o', error );
+				this.props.errorNotice( 'Danger!!!' );
```
Comment (sirreal) [inline client/jetpack-connect/signup.js]: Do something better. Errors were not handled before 😱

**Comment by sirreal** [comment]
This is working with basic signups. Not finished yet, but worth looking at.

There are pieces of redux state that appear to be reused in several auth steps. I'm tempted to pull those into this component where it can track what's happening in local state.

**Comment by sirreal** [comment]
Problems:

- Signupform has `submitting` prop and stateful prop. Apparently no way to tell the signupform that it's not submitting anymore 🤷‍♂️  https://github.com/Automattic/wp-calypso/blob/31ef652b9970f8980a74adaa2b326138ada3546e/client/components/signup-form/index.jsx#L92
- Notices aren't rendered (no `GlobalNotices` in the logged out layout): https://github.com/Automattic/wp-calypso/blob/master/client/layout/logged-out.jsx

These are not new issues but existing limitations. We should consider moving forward with this, which will allow for work on #23483 to continue following this pattern.

File: client/state/jetpack-connect/actions.js [hunk 2]
```diff
@@ -197,28 +196,34 @@ export function retryAuth( url, attemptNumber ) {
...
-			dispatch( {
-				type: JETPACK_CONNECT_CREATE_ACCOUNT_RECEIVE,
-				data,
-				error,
-				userData,
-			} );
-		} );
+		try {
+			const data = await wpcom.undocumented().usersNew( userData );
+			const bearerToken = makeParser(
```
Comment (sirreal) [inline client/state/jetpack-connect/actions.js]: `makeParser` was added for data-layer, but it's handy for more things like this. I'd like to "promote" it to a framework level util (`client/lib/make-json-schema-parser`).

File: client/state/jetpack-connect/test/actions.js [hunk 3]
```diff
@@ -410,53 +408,29 @@ describe( '#createAccount()', () => {
...
-			},
-		} ) );
-
-		const spy = jest.fn();
-		createAccount()( spy );
-		expect( spy ).toHaveBeenCalledWith( { type: JETPACK_CONNECT_CREATE_ACCOUNT } );
-	} );
-
-	test( 'should dispatch receive action with appropriate data', () => {
+	test( 'should resolve with the bearer token', async () => {
```
Comment (sirreal) [inline client/state/jetpack-connect/test/actions.js]: This `async` can probably be removed

File: client/jetpack-connect/signup.js [hunk 4]
```diff
@@ -150,22 +158,17 @@ export class JetpackSignup extends Component {
 						) }
 						submitButtonText={ this.props.translate( 'Create your account' ) }
 						submitForm={ this.handleSubmitSignup }
-						submitting={ isAuthorizing }
-						suggestedUsername={ get( userData, 'username', '' ) }
```
[the remark this answers is not in the document]
Comment (sirreal) [reply client/jetpack-connect/signup.js]: That's my understanding. It's also a required prop on `SignupForm`, so they default `''` is probably the only relevant value, and what we maintain with this PR.

File: client/jetpack-connect/signup.js [hunk 5]
```diff
@@ -76,31 +73,45 @@ export class JetpackSignup extends Component {
...
+						this.setState( {
+							newUsername: userData.username,
+							bearerToken,
+							isCreatingAccount: false,
+						} ),
+					error => {
+						debug( 'Signup error: %o', error );
+						this.setState( { isCreatingAccount: false } );
+						this.props.errorNotice(
+							this.props.translate( 'There was a problem creating your account.' )
```
Comment (seear) [inline client/jetpack-connect/signup.js]: "...Please _contact support._" ?

**Comment by seear** [comment]
> Verify that master fails in the same state as this branch: dead form, no notices

Mm, not too slick.

Also noting here that (in production) the button remains enabled when there are errors on the input fields. Clicking the button doesn't seem to do anything.

screen shot 2018-04-06 at 15 02 31

**Comment by sirreal** [comment]
I'm going to push another commit before string freeze to update translation. Note that e2e [is passing](https://circleci.com/gh/Automattic/wp-e2e-tests-for-branches/4500) on 14bfa6e

File: client/jetpack-connect/signup.js (hunk 5 above)
Comment (sirreal) [reply client/jetpack-connect/signup.js]: 👍 aa66c9004d5ce530ff9c47b9ad53a0bd50ca00bf

**Comment by sirreal** [comment]
Thanks for the thorough review, @seear!

**Pull request merged**